src/components/application_manager/test/rpc_service_impl_test.cc

```diff
@@ -301,12 +331,362 @@ TEST_F(RPCServiceImplTest,
+  message_to_send->set_function_id(kFunctionId);
+  message_to_send->set_message_type(static_cast<am::MessageType>(2));
+  message_to_send->set_json_message(std::string("message"));
+  am::BinaryData binaryData({0, 0, 0});
+  message_to_send->set_binary_data(&binaryData);
+
+  NiceMock<test::components::connection_handler_test::MockConnectionHandler>
+      mock_conn_handler;
+  EXPECT_CALL(mock_app_mngr_, connection_handler())
+      .WillRepeatedly(ReturnRef(mock_conn_handler));
+  ON_CALL(mock_conn_handler, GetConnectionSessionsCount(_))
+      .WillByDefault(Return(2));
+  EXPECT_CALL(mock_protocol_handler_, IsRPCServiceSecure(_))
+      .WillOnce(Return(true));
+  ON_CALL(mock_app_mngr_, application(_)).WillByDefault(Return(nullptr));
+  EXPECT_CALL(*mock_rpc_protection_manager_, CheckPolicyEncryptionFlag(_, _, _))
+      .WillOnce(Return(true));
+  EXPECT_CALL(mock_protocol_handler_, SendMessageToMobileApp(_, _, _))
+      .WillOnce(Return());
+  EXPECT_CALL(*mock_rpc_protection_manager_,
+              RemoveFromEncryptionNeededCache(_, _))
+      .WillOnce(Return());
+  EXPECT_CALL(mock_conn_handler, CloseSession(_, _)).WillOnce(Return());
+  message_handler->Handle(
+      am::rpc_service::impl::MessageToMobile(message_to_send, true));
+}
+
+TEST_F(RPCServiceImplTest,
+       SendMessageToMobile_InvalidMessageType_ConvertSOtoMessageReturnsFalse) {
+  auto message = CreateMessage(smart_objects::SmartType_Invalid);
+  (*message)[am::strings::msg_params][am::strings::result_code] =
+      ns_smart_device_link_rpc::V1::Result::SUCCESS;
+  MockAppPtr = std::make_shared<NiceMock<MockApplication> >();
+  ON_CALL(mock_app_mngr_, application(_)).WillByDefault(Return(MockAppPtr));
+  ON_CALL(*MockAppPtr, protocol_version())
+      .WillByDefault(
+          Return(protocol_handler::MajorProtocolVersion::PROTOCOL_VERSION_1));
+  // rpc_passing part, expect false
+  ON_CALL(mock_app_mngr_, GetAppServiceManager())
+      .WillByDefault(ReturnRef(mock_app_service_nmgr_));
+  auto rpc_pass_handler =
+      new am::RPCPassingHandler(mock_app_service_nmgr_, mock_app_mngr_);
+  ON_CALL(mock_app_service_nmgr_, GetRPCPassingHandler())
+      .WillByDefault(ReturnRef(*rpc_pass_handler));
+  ON_CALL(mock_app_service_nmgr_, GetActiveServices())
+      .WillByDefault(Return(std::vector<am::AppService>()));
+  ON_CALL(mock_app_mngr_, GetPluginManager())
+      .WillByDefault(ReturnRef(mock_rpc_plugin_manager_));
+  utils::Optional<am::plugin_manager::RPCPlugin> mock_rpc_plugin_opt(
+      mock_rpc_plugin_);
+  ON_CALL(mock_rpc_plugin_manager_, FindPluginToProcess(_, _))
+      .WillByDefault(Return(mock_rpc_plugin_opt));
+  ON_CALL(mock_rpc_plugin_, PluginName()).WillByDefault(Return(""));
+
+  EXPECT_CALL(*MockAppPtr, protocol_version())
+      .WillOnce(
+          Return(protocol_handler::MajorProtocolVersion::PROTOCOL_VERSION_1));
+  rpc_service_->SendMessageToMobile(message);
+  delete rpc_pass_handler;
```

Comment:
Alocate it in stack to avoid deletion

Reply:
Unfortunately, I can't allocate instance of this class on stack because of synchronization issues. As I see from the older code, only this way to create is working.